### PR TITLE
Proper individual dicts

### PR DESCRIPTION
**Pull request by Dranero**
Still in progress

*Added Future based mp handler
*Consolidated multipreseesing tasks
*Refactored seed usage
*Added proper genom dicts for layered Brains

TODO
*Messure speedup in comparison to master

### CODE REVIEW

**Comment by Dranero** [comment]
The Speedup options speed up an <details>
  <summary>Experiment</summary>
{
  "environment": "ReacherMemory-v0",
  "random_seed": 1,
  "number_generations": 100,
  "optimizer": {
    "type": "CMA_ES",
    "population_size": 100,
    "sigma": 1.0,
    "checkpoint_frequency": 0,
    "hof_size": 5,
    "mu": 25,
    "efficiency_weight": 0.0,
    "fix_seed": "false"
  },
  "brain": {
    "type": "ELMAN",
    "hidden_layer_structure": [10],
    "diagonal_hidden_to_hidden": true,
    "use_bias": true,
    "optimize_initial_neuron_state": false
  },
  "episode_runner": {
    "number_fitness_runs": 25,
    "environment_attributes": {
      "type": "ReacherMemoryAttr",
      "observation_frames": 20,
      "memory_frames": 20,
      "action_frames": 50
    }
  }
}

</details> up by ~3%

File: neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py [hunk 1]
```diff
@@ -65,35 +68,82 @@ def fct_mutation_learned(ind1):
...
+                                          logbook=self.logbook, last_seed=current_seed, strategy=None,
+                                          recorded_individuals=self.recorded_individuals)
+        return self.logbook
+
+    @classmethod
+    def strip_strategy_from_population(cls, population, strategy_parameter_per_gene=False):
+        if len(population) == 0:
+            return population
+        return list(np.array(population)[:,
+                    :-(cls.get_number_genes_for_strategy(len(population[0]), strategy_parameter_per_gene))])
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py:141]: Die Methode ist ziemlich unübersichtlich, was u.A. daran liegen dürfte, dass meine ursprüngliche Implementierung davon eher mangelhaft war

Der Wert für `strategy_parameter_per_gene` ist für den optimizer immer gleich, oder? Man könnte den zu einer member variable machen. Das gleiche für `len(population[0]`. Und dann liefert `get_number_genes_for_strategy` auch immer den gleichen wert zurück, oder? Dann kann man das anstatt zu einer methode auch zu eine Variable machen.

Der Fall mit  `len(population) == 0` trifft nur im Fehlerfall ein, oder? Wenn ja, dann  entweder eine Exception werfen, oder das `if` ganz  weglassen (und darauf vertrauen, dass der fehler dann irgendwo anders geworfen wird). Dieses stillschweigende Fehler-Beheben ist auf jeden Fall immer schlecht. Weil wenn der Fehler dann auftritt, dann wird das experiment sowieso ganz gegen die wand laufen. Und wenn dann viele solche stellen drin sind, wo stillschweigende Fehler "gefixt" werden, dann wird das troubleshooting viel viel schwieriger. 



```suggestion
    def strip_strategy_from_population(self, population):
        return list(np.array(population)[:, :-self.number_strategy_params_in_genome])
```

File: neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py [hunk 2]
```diff
@@ -65,35 +68,82 @@ def fct_mutation_learned(ind1):
...
+
+    def get_number_genes_for_strategy(self, individual_size):
+        return individual_size // 2 if self.conf.strategy_parameter_per_gene else 2
+
+    @classmethod
+    def strip_strategy_from_individual(cls, individual, strategy_parameter_per_gene=False):
+        return list(
+            np.array(individual)[:-(cls.get_number_genes_for_strategy(len(individual), strategy_parameter_per_gene))])
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py:149]: DRY. Kann die Methode vielleicht sogar ganz entfernt werden?

```suggestion
    def strip_strategy_from_individual(self, individual):
        return self.strip_strategy_from_population([individual])[0]
```

File: neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py [hunk 3]
```diff
@@ -20,21 +19,25 @@ class OptimizerMuPlusLambda(IOptimizer[OptimizerMuLambdaCfg]):
...
-                                                    from_checkpoint)
-        toolbox = self.toolbox
+        super(OptimizerMuPlusLambda, self).__init__(eval_fitness, random_seed, conf, stats, map_func,
+                                                    from_checkpoint, reset_hof)
 
         if self.conf.strategy_parameter_per_gene:
             individual_size *= 2
         else:
             # add two genes for strategy parameters used in mutate
             individual_size += 2
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/optimizer/optimizer_mu_lambda.py:29]: Siehe mein kommentar bei `strip_strategy_from_population`

```suggestion
        if self.conf.strategy_parameter_per_gene:
            # add a gene as strategy parameters for each individual actual gene
            self.number_strategy_params_in_genome = individual_size 
        else:
            # add two genes as strategy parameters for the entire genome
            self.number_strategy_params_in_genome += 2 
        individual_size += self.number_strategy_params_in_genome
```

File: neuro_evolution_ctrnn/brains/i_pytorch_brain.py [hunk 4]
```diff
@@ -13,7 +13,7 @@
 PytorchBrainConfigClass = TypeVar('PytorchBrainConfigClass', bound=IPytorchBrainCfg)
 
 
-class IPytorchBrain(nn.Module, IBrain[PytorchBrainConfigClass], abc.ABC):
+class IPytorchBrain(nn.Module, IBrain, abc.ABC, Generic[PytorchBrainConfigClass]):
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/brains/i_pytorch_brain.py:16]: Was macht das `Generic`?

File: neuro_evolution_ctrnn/optimizer/i_optimizer.py [hunk 5]
```diff
@@ -140,3 +123,116 @@ def shape_fitness_weighted_ranks(self, population):
...
+                raise NotImplementedError()
+        else:
+            seed_for_generation = 0
+            seeds_for_evaluation = np.zeros(len(candidates), dtype=np.int64)
+            seeds_for_recorded = np.zeros(len(self.recorded_individuals), dtype=np.int64)
+
+        brain_genomes_recorded = self.strip_strategy_from_population(self.recorded_individuals)
+        nevals = len(candidates) + len(brain_genomes_recorded)
+
+        novelty_data = (
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/optimizer/i_optimizer.py:154]: alles was mit novelty zu tun hat, sollte mMn ersatzlos entfernt werden. Also insb. `seeds_for_recorded`, `brain_genomes_recorded`, `novelty_data` und `self.conf.novelty`. 

(Sollte jedoch in einem Separatem PR geschehen. Würde ich "demnächst" auh selbst machen)

File: neuro_evolution_ctrnn/tools/episode_runner.py [hunk 6]
```diff
@@ -90,10 +94,38 @@ def eval_fitness(self, individual, seed, render: bool = False, record: str = Non
             fitness_total += fitness_current
             steps_total += t
 
-        compressed_behavior = None
-        if hasattr(env, "get_compressed_behavior"):
-            # 'get_compressed_behavior' exists if any wrapper is a BehaviorWrapper
-            assert callable(env.get_compressed_behavior)
-            compressed_behavior = env.get_compressed_behavior()
+        # TODO does this gets used anywhere?
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/tools/episode_runner.py:97]: > does this gets used anywhere?

Ja, für Novelty-search, wenn man die novelty aus den internen zustand des CTRNN ableiten will. Kann aber alles entfernt werden. Die idee war, dass das für memory-experiment interessant werden könnte

File: neuro_evolution_ctrnn/tools/experiment.py [hunk 7]
```diff
@@ -52,7 +53,7 @@ def _setup(self):
         # note: the environment defined here is only used to initialize other classes, but the
         # actual simulation will happen on freshly created local  environments on the episode runners
         # to avoid concurrency problems that would arise from a shared global state
-        self.env_template = env
+        EpisodeRunner._env = self.env_template = env
```
Comment (bjuergens) [inline neuro_evolution_ctrnn/tools/experiment.py:56]: Stimmt der Kommentar oben drüber noch, wenn hier die env als Klassen-Variable gespeichert wird?

**Review by bjuergens — APPROVED** [review]
Gefällt mir sehr gut. Kann von mir aus sofort gemergt werden. 

Es gibt ein paar kleinigkeiten, die man noch machen könnte, die ich aber auch selbst in einem nachfolge-PR machen kann.
